main.py

```diff
@@ -472,11 +477,52 @@ def update_script():
             f.write(r.text)
 
 
+def set_up_plugins(arg_array):
+    if not get_plugin_dir() and not os.path.exists('plugins/') and '-z' not in arg_array:
+        arg_array.insert(0, '-z')
+    elif get_plugin_dir():
+        pl_dir = get_plugin_dir()
+        pl_dir = f'{pl_dir}' if not pl_dir.endswith('/') or not pl_dir.endswith('\\') else pl_dir
+        if not os.path.exists(f'{pl_dir}__init__.py'):
+            with open(f'{pl_dir}__init__.py', 'w') as f:
+                f.write(0)
+
+    return arg_array
+
+
 def main():
+    arg_array = sys.argv[1::]
+
+    read_config()
+    arg_array = set_up_plugins(arg_array)
+    plugins = [
+        getattr(__import__(f'plugins.{item.replace(".py", "")}'), item.replace('.py', ''))
+        for item in os.listdir(get_plugin_dir() or 'plugins')
+        if '__init__.py' not in item and '__pycache__' not in item and item.endswith('.py')
+    ]
+    parser = argparse.ArgumentParser()
+
+    if '-z' not in arg_array and '--noplugins' not in arg_array:
+        for plugin in plugins:
+            try:
+                plugin.pre_argparse(parser)
+            except:
+                pass
+
     if has_version_difference():
         print("Der er en opdatering til STS. Hent den fra GitHub eller kør sts med --update.", file=sys.stderr)
 
-    args = parse(sys.argv[1::])
+    args = parse(arg_array, parser)
+
+    if args.user is None:
+        args.user = get_saved_user()
+
+    if not args.noplugins:
+        for plugin in plugins:
+            try:
+                plugin.run(wares, args, arg_array, parser, SHORTHANDS)
+            except:
+                pass
```

Comment:
Could be nice to emit exceptions to user when running third-party plugins. Maybe print caught exception?

Reply:
Resolved in abe2943